accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/DecoratedAccessioningService.java

```diff
@@ -0,0 +1,143 @@
+    }
+
+    private List<DB_ACCESSION> parse(List<ACCESSION> accessions) {
+        return accessions.stream().map(parsingFunction).filter(Objects::nonNull).collect(Collectors.toList());
+    }
+
+    @Override
+    public AccessionWrapper<MODEL, HASH, ACCESSION> getByAccessionAndVersion(ACCESSION accession, int version)
+            throws AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        return decorate(service.getByAccessionAndVersion(parse(accession), version));
+    }
+
+    private DB_ACCESSION parse(ACCESSION accession) throws AccessionDoesNotExistException {
+        DB_ACCESSION dbAccession = parsingFunction.apply(accession);
+        if(dbAccession ==null){
+            throw new AccessionDoesNotExistException(accession);
+        }
+        return dbAccession;
+
+    }
+
+    @Override
+    public AccessionVersionsWrapper<MODEL, HASH, ACCESSION> update(ACCESSION accession, int version, MODEL message)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        return decorate(service.update(parse(accession), version, message));
+    }
+
+    private AccessionVersionsWrapper<MODEL, HASH, ACCESSION> decorate(
+            AccessionVersionsWrapper<MODEL, HASH, DB_ACCESSION> accessionVersions) {
+        return new AccessionVersionsWrapper<>(decorate(accessionVersions.getModelWrappers()));
+    }
+
+    @Override
+    public AccessionVersionsWrapper<MODEL, HASH, ACCESSION> patch(ACCESSION accession, MODEL message)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        return decorate(service.patch(parse(accession), message));
+    }
+
+    @Override
+    public void deprecate(ACCESSION accession, String reason) throws AccessionMergedException,
+            AccessionDoesNotExistException, AccessionDeprecatedException {
+        service.deprecate(parse(accession), reason);
+    }
+
+    @Override
+    public void merge(ACCESSION accessionOrigin, ACCESSION mergeInto, String reason) throws AccessionMergedException,
+            AccessionDoesNotExistException, AccessionDeprecatedException {
+        service.merge(parse(accessionOrigin), parse(mergeInto), reason);
+    }
+
+    public static <MODEL, HASH, DB_ACCESSION> DecoratedAccessioningService<MODEL, HASH, DB_ACCESSION, String>
+    buildPrefixAccessionService(AccessioningService<MODEL, HASH, DB_ACCESSION> service, String prefix,
+                                Function<String, DB_ACCESSION> parseFunction) {
+        return new DecoratedAccessioningService<>(service, accession -> prefix + accession,
+                s -> {
+                    if (!Objects.equals(s.substring(0, prefix.length()), prefix)) {
+                        return null;
+                    }
+                    return parseFunction.apply(s.substring(prefix.length()));
```

Comment:
I don't think you need `parseFunction` as an argument for this method. If `decorated_accession = prefix + db_accession`, then surely `db_accession = decorated_accession - prefix`? What else can `parseFunction` do here?

Reply:
The parse function literaly parses or transforms the string part that represents the underlying accesssion to the database type used.

Reply:
I see. So it does type conversion if needed. That's fine then.